### PR TITLE
'Solution'

### PR DESCRIPTION
**Pull request by tarasulo**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,52 +1,98 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List</p>
  */
 public class ArrayList<T> implements List<T> {
+    private int size;
+    private int defaultCapacyty = 10;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: better make it as constant

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,52 +1,98 @@
...
+    private Object[] arr;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: arr?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,52 +1,98 @@
...
+    public ArrayList(int currentCapacity) {
+        if (currentCapacity > 0) {
+            this.arr = new Object[currentCapacity];
+        } else {
+            throw new IllegalArgumentException("Wrong initial capacity: " + currentCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this.arr = new Object[defaultCapacyty];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        arr = new Object[defaultCapacyty];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,52 +1,98 @@
...
-
+        setCapacity();
+        arr[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        setCapacity();
+        if (index >= size) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: if index <0 ?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,52 +1,98 @@
...
+        }
+        System.arraycopy(arr, index, arr, index + 1, size - index);
+        arr[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: optional:
see how this method is implemented in the source code

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,52 +1,98 @@
...
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: if index <0 or < size?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,52 +1,98 @@
...
+                return remove(i);
+            }
+        }
+        throw new ArrayIndexOutOfBoundsException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return this.size;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        return size;
```

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 2 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: chenged

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: agree

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: saw

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: agree

File: src/main/java/core/basesyntax/ArrayList.java (hunk 7 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,52 +1,101 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List</p>
  */
 public class ArrayList<T> implements List<T> {
+    private int size;
+    private final int defaultCapacyty = 10;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: https://mate-academy.github.io/style-guides/java/java.html#s5.2.4-constant-names

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,52 +1,101 @@
...
+    private Object[] arrList;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: rename to dataElements

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,52 +1,101 @@
...
+
+    public ArrayList(int currentCapacity) {
+        if (currentCapacity > 0) {
+            this.arrList = new Object[currentCapacity];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: remove `this` everuthere

File: src/main/java/core/basesyntax/ArrayList.java (hunk 8 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 9 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

File: src/main/java/core/basesyntax/ArrayList.java (hunk 10 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: done

**Comment by tarasulo** [comment]
[ERROR] /home/travis/build/mate-academy/jv-array-list/src/main/java/core/basesyntax/ArrayList.java:10: Abbreviation in name 'DEFAULT_CAPACITY' must contain no more than '2' consecutive capital letters. [AbbreviationAsWordInName]   так помилка

File: src/main/java/core/basesyntax/ArrayList.java (hunk 8 above)
Comment (tarasulo) [reply src/main/java/core/basesyntax/ArrayList.java]: [ERROR] /home/travis/build/mate-academy/jv-array-list/src/main/java/core/basesyntax/ArrayList.java:10: Abbreviation in name 'DEFAULT_CAPACITY' must contain no more than '2' consecutive capital letters. [AbbreviationAsWordInName]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,52 +1,101 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List</p>
  */
 public class ArrayList<T> implements List<T> {
+    private int size;
+    private final int defaultCapacity = 10;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
    private static final int DEFAULT_CAPACITY = 10;
```

**Pull request closed without merging**